sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java

```diff
@@ -0,0 +1,551 @@
+      return toBuilder().setDeduplicateRecords(deduplicateRecords).build();
+    }
+
+    /**
+     * Set a factory that creates a {@link org.apache.beam.sdk.io.solace.broker.SempClientFactory}.
+     *
+     * <p>The factory `create()` method is invoked in each instance of an {@link
+     * org.apache.beam.sdk.io.solace.read.UnboundedSolaceReader}. Created {@link
+     * org.apache.beam.sdk.io.solace.broker.SempClient} has to communicate with broker management
+     * API. It must support operations such as:
+     *
+     * <ul>
+     *   <li>query for outstanding backlog bytes in a Queue,
+     *   <li>query for metadata such as access-type of a Queue,
+     *   <li>requesting creation of new Queues.
+     * </ul>
+     *
+     * <p>An existing implementation of the SempClientFactory includes {@link
+     * org.apache.beam.sdk.io.solace.broker.BasicAuthSempClientFactory} which implements the Basic
+     * Authentication to Solace.
+     *
+     * <p>To use it, specify the credentials with the builder methods.
+     *
+     * <p>The format of the host is `[Protocol://]Host[:Port]`
+     *
+     * <pre>{@code
+     * .withSempClientFactory(
+     *         BasicAuthSempClientFactory.builder()
+     *                 .host("http://" + options.getSolaceHost() + ":8080")
+     *                 .username(options.getSolaceUsername())
+     *                 .password(options.getSolacePassword())
+     *                 .vpnName(options.getSolaceVpnName())
+     *                 .build())
+     * }</pre>
+     */
+    public Read<T> withSempClientFactory(SempClientFactory sempClientFactory) {
+      checkState(sempClientFactory != null, "SolaceIO.Read: sempClientFactory must not be null.");
+      return toBuilder().setSempClientFactory(sempClientFactory).build();
+    }
+
+    /**
+     * Set a factory that creates a {@link SessionService}.
+     *
+     * <p>The factory `create()` method is invoked in each instance of an {@link
+     * org.apache.beam.sdk.io.solace.read.UnboundedSolaceReader}. Created {@link SessionService} has
+     * to be able to:
+     *
+     * <ul>
+     *   <li>initialize a connection with the broker,
+     *   <li>check liveliness of the connection,
+     *   <li>close the connection,
+     *   <li>create a {@link org.apache.beam.sdk.io.solace.broker.MessageReceiver}.
+     * </ul>
+     *
+     * <p>An existing implementation of the SempClientFactory includes {@link
+     * org.apache.beam.sdk.io.solace.broker.BasicAuthJcsmpSessionService} which implements the Basic
+     * Authentication to Solace. *
+     *
+     * <p>To use it, specify the credentials with the builder methods. *
+     *
+     * <p>The format of the host is `[Protocol://]Host[:Port]` *
+     *
+     * <pre>{@code
+     * BasicAuthJcsmpSessionServiceFactory.builder()
+     *         .host(options.getSolaceHost())
+     *         .username(options.getSolaceUsername())
+     *         .password(options.getSolacePassword())
```

Comment:
I had trouble finding where the options class was. Is it an extension of PipelineOptions?

Reply:
Actually, options class will not be a part of this (or any) PR. A user can create their own custom PipelineOptions. I will replace the references to the option fields with placeholders.